src/editor/components/scenegraph/SceneGraph.js

```diff
@@ -62,307 +51,63 @@ export default class SceneGraph extends React.Component {
-      nextIter++
-    ) {
-      const nextEl = this.state.entities[nextIter].entity;
-      if (this.isVisibleInSceneGraph(nextEl)) {
-        return nextIter;
+      if (!this.shouldRenderNode(node)) {
+        continue;
       }
-    }
-    return -1;
-  };
-
-  onChangeFilter = (evt) => {
-    const filter = evt.target.value;
-    this.setState({ filter: filter });
-    this.updateFilteredEntities(filter);
-  };
-
-  updateFilteredEntities(filter) {
-    this.setState({
-      filteredEntities: this.getFilteredEntities(filter)
-    });
-  }
-
-  clearFilter = () => {
-    this.setState({ filter: '' });
-    this.updateFilteredEntities('');
-  };
 
-  toggleLeftBar = () => {
-    this.setState({ leftBarHide: !this.state.leftBarHide });
-  };
-
-  renderEntities = () => {
-    let entityOptions = this.state.filteredEntities.filter((entityOption) => {
-      if (
-        !this.isVisibleInSceneGraph(entityOption.entity) &&
-        !this.state.filter
-      ) {
-        return false;
-      } else {
-        return true;
-      }
-    });
-
-    // wrap entities of layer level 1 in <div class="layer">
-    let layerEntities = [];
-    let resultEntities = [];
-    // let activeLayer = false;
-    for (let i = 1; i < entityOptions.length; i++) {
-      const entityOption = entityOptions[i];
-      const entity = (
-        <Entity
-          {...entityOption}
-          key={i}
-          isFiltering={!!this.state.filter}
-          isExpanded={this.isExpanded(entityOption.entity)}
-          isSelected={this.props.selectedEntity === entityOption.entity}
-          selectEntity={this.selectEntity}
-          toggleExpandedCollapsed={this.toggleExpandedCollapsed}
-          isInitiallyExpanded={this.state.initiallyExpandedEntities.some(
-            (item) => item === entityOption.entity.id
-          )}
-          initiallyExpandEntity={() => {
-            this.toggleExpandedCollapsed(entityOption.entity);
-            this.setState((prevState) => ({
-              ...prevState,
-              initiallyExpandedEntities: [
-                ...prevState.initiallyExpandedEntities.filter(
-                  (item) => item !== entityOption.entity.id
-                ),
-                entityOption.entity.id
-              ]
-            }));
-          }}
-        />
-      );
-      layerEntities.push(entity);
-      if (i === entityOptions.length - 1 || entityOptions[i + 1].depth === 1) {
-        const className = classNames({
-          layer: true,
-          active: layerEntities[0].props.isSelected
-        });
-        resultEntities.push(
-          <div className={className} key={i}>
-            {layerEntities}
+      const children = node.children || [];
+      const { renderedChildren, isExpanded: childIsExpanded } =
+        this.renderSceneNodes(Array.from(children), depth + 1);
+
+      const isSelected = node === this.props.selectedEntity;
+      isExpanded = isExpanded || isSelected || childIsExpanded;
```

Comment:
im figuring out whether or not the node should be expanded based on the a selected child node in here. There seems to be an issue where after a close of an expanded node and then selection on the editor, it doesn't seem to re-expand